### ISSUE TITLE
Crash when removing folders in sublime text

### ISSUE THREAD

**Issue by gerardroche** [phpactor/phpactor#2516]
phpactor crashes including crashing other things like firefox, terminal, and also gnome shell itself. Obviously this is quiet a severe crash. 

I'm using phpactor in [Sublime Text LSP](https://github.com/sublimelsp/LSP) and can consistently reproduce the crash when removing several folders in quick succession. It's not the only context in which the crash occurs, but I can reproduce that one.

I reported the issue in the Sublime text [LSP plugin](https://github.com/sublimelsp/LSP/issues/2380), but I've narrowed the issue down to phpactor. There are additional details on the issue, but I'll add some here.

My Sublime Text lsp config, the configurations outside of "clients" don't seem to be relevant, but I'll show anyway:

```
{

    "diagnostics_delay_ms": 50,
    "diagnostics_gutter_marker": "dot",
    "log_debug": false,
    "log_server": false,
    "show_code_actions": "",
    "show_diagnostics_count_in_view_status": false,
    "show_diagnostics_in_view_status": false,
    "show_inlay_hints": false,
    "show_multiline_diagnostics_highlights": false,
    "show_view_status": false,

    "diagnostics_highlight_style": {
        "error": "",
        "warning": "",
        "info": "stippled",
        "hint": "stippled"
    },

    "clients": {
        "phpactor": {
            "command": [
                "php",
                "-d",
                "xdebug.mode=off",
                "/usr/local/bin/phpactor",
                "language-server",
            ],
            "enabled": true,
            "languageId": "php",
            "scopes": [
                "source.php",
                "embedding.php"
            ],
            "syntaxes": [
                "Packages/PHP/PHP.sublime-syntax"
            ],
        },
    },
}
```

I don't think it probably matters how I have it installed, but I have phpactor installed via composer. The `/usr/local/bin/phpactor` is symlinked to `bin/phpactor` in my cloned phpactor repository which is installed, for example, like this':

```
cd ~
git clone https://github.com/phpactor/phpactor
cd phpactor
composer install --no-dev --optimize-autoloader
cd /usr/local/bin
sudo ln -s ~/phpactor/phpactor/bin/phpactor
``` 

## To reproduce the issue

1. Add several project folders to sublime 
2. Open a php file (opening a php file initialises the phpactor language-server).
3. Remove two of the folders one after the other quickly (less than about 1 sec apart)

I use my [Sesame](https://packagecontrol.io/packages/Sesame) plugin so it's easy to remove folders quickly, but you can do it without Sesame, you just need to be quick with you mouse.

I enabled [phpactor logging](https://phpactor.readthedocs.io/en/master/development/debugging.html#logging) but there is nothing of note in the log.

**Other notes**

- Sublime 4169
- Ubuntu 22.04
- Gnome Shell 42.9
- Windowing System Wayland

In Sublime the code to remove a folder is:

```
def _remove_folder(window: Window, folder: str) -> None:
    window.run_command('remove_folder', {
        'dirs': [folder]
    })
```

So you should be able to run the command via the console if you like:

```
>>> sublime.active_window().run_command('remove_folder', {'dirs': ['/path/to/folder']})
```

Otherwise right click the folder in side bar and select Remove Folder from Project.

**Comment by gerardroche** [comment]
See https://github.com/sublimelsp/LSP/issues/2380#issuecomment-1907192013

I think it's this: 

https://github.com/amphp/process/blob/v1.1.4/lib/Internal/Posix/Runner.php#L197

```
@\posix_kill($pid, 9);
```

When there is an error, `$pid` is null and this causes everything to crash. When add a null check, the problem seems to go away. 

The [php docs](https://www.php.net/posix-kill) says nothing about null, but I think a null would be coerced to 0 so `posix_kill()` would receive 0 in the case of an error.

```
php > var_dump((int) null);
int(0)
```

phpactor is using an old version of amphp/process: v1.1.4 - the newest version is v2.0.1 so maybe it's fixed. The newer code looks a lot better.

**Comment by gerardroche** [comment]
This is an issue in amphp/process dependency. I've sent a patch:

https://github.com/amphp/process/pull/68

When it's merged, the dependency can be updated to get the patch.

**Comment by dantleech** [comment]
Neovim sometimes crashes when deleting folders, I wonder if it can be related.

Upgrading to Amp 2.0 would be great in general, but also they switch to fibers, so not sure how much work it would be.


The only place in the code I can find that directly invokes `kill` is https://github.com/phpactor/phpactor/blob/fa0413d33093cffd0c52ef5312ac2d2157e760c2/lib/Amp/Process/ProcessUtil.php#L23 which only gets called when killing Psalm or the Phpactor diagnostics process after a timeout.

Did you identify the stack trace for when this happens in Sublime?

**Comment by gerardroche** [comment]
I confirmed that it was the `posix_kill()` calls in https://github.com/amphp/process/pull/68/files. Essentially, if the pid is null, `posix_kill()` kills most open programs. I think it might be technically something like, all programs in the same process group, or sibling processes. Either way, it kills a lot of programs.

I could never discover why the pid is null. I assume there is an error somewhere, but there is no exception. I didn't get around to trying to use debugger to see what is happening.

I also attempted to upgrade to Amp 2.0, but it's a little more complicated than I anticipated, so backed out of it. The issue might be fixed in v2. I seen related commits in the git log. Even if the issue is still v2, it's an easy null-check fix.
